File: python/mirage/commands/builtin/nextcloud/file.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

from functools import partial

from mirage.accessor.nextcloud import NextcloudAccessor
from mirage.cache.index import IndexCacheStore
from mirage.commands.builtin.generic.file import file_cmd as generic_file
from mirage.commands.registry import command
from mirage.commands.spec import SPECS
from mirage.core.nextcloud.glob import resolve_glob
from mirage.core.nextcloud.read import read_bytes
from mirage.core.nextcloud.stat import stat
from mirage.io.types import ByteSource, IOResult
from mirage.types import PathSpec


@command("file", resource="nextcloud", spec=SPECS["file"])
async def file(
    accessor: NextcloudAccessor,
    paths: list[PathSpec],
    *texts: str,
    stdin: bytes | None = None,
    b: bool = False,
    i: bool = False,
    index: IndexCacheStore = None,
    **_extra: object,
) -> tuple[ByteSource | None, IOResult]:
    if not paths:
        raise ValueError("file: missing operand")
    paths = await resolve_glob(accessor, paths, index)
    return await generic_file(paths,
                              read_bytes=partial(read_bytes, index=index),
                              stat_fn=partial(stat, index=index),
                              accessor=accessor,
                              b=b,
                              i=i)

File: typescript/packages/node/src/commands/builtin/ssh/file/file.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import {
  FileType,
  IOResult,
  ResourceName,
  command,
  detectFileType,
  formatFileResult,
  specOf,
  type ByteSource,
  type CommandFnResult,
  type CommandOpts,
  type PathSpec,
  formatRecords,
} from '@struktoai/mirage-core'
import { stat as sshStat } from '../../../../core/ssh/stat.ts'
import { read as sshRead } from '../../../../core/ssh/read.ts'
import type { SSHAccessor } from '../../../../accessor/ssh.ts'

const ENC = new TextEncoder()

async function fileCommand(
  accessor: SSHAccessor,
  paths: PathSpec[],
  texts: string[],
  opts: CommandOpts,
): Promise<CommandFnResult> {
  if (paths.length === 0) {
    return [null, new IOResult({ exitCode: 1, stderr: ENC.encode('file: missing operand\n') })]
  }
  const brief = opts.flags.b === true
  const mime = opts.flags.i === true
  const lines: string[] = []
  for (const p of paths) {
    const s = await sshStat(accessor, p)
    if (s.type === FileType.DIRECTORY) {
      lines.push(formatFileResult(p.original, FileType.DIRECTORY, brief, mime))
      continue
    }
    let header: Uint8Array
    try {
      const raw = await sshRead(accessor, p)
      header = raw.subarray(0, 512)
    } catch {
      header = new Uint8Array(0)
    }
    const result = detectFileType(header, s)
    lines.push(formatFileResult(p.original, result, brief, mime))
  }
  const out: ByteSource = formatRecords(lines)
  return [out, new IOResult()]
}

export const SSH_FILE = command({
  name: 'file',
  resource: ResourceName.SSH,
  spec: specOf('file'),
  fn: fileCommand,
})
